core/src/dht/dht_reducers.rs

```diff
@@ -0,0 +1,144 @@
+
+use action::{Action, ActionWrapper};
+use cas::{content::AddressableContent, storage::ContentAddressableStorage};
+use context::Context;
+use dht::dht_store::DhtStore;
+use eav::EntityAttributeValueStorage;
+use hash_table::entry::Entry;
+use std::sync::Arc;
+
+// A function that might return a mutated DhtStore
+type DhtReducer<CAS, EAVS> =
+    fn(Arc<Context>, &DhtStore<CAS, EAVS>, &ActionWrapper) -> Option<DhtStore<CAS, EAVS>>;
+
+/// DHT state-slice Reduce entry point.
+/// Note: Can't block when dispatching action here because we are inside the reduce's mutex
+pub fn reduce<CAS, EAVS>(
+    context: Arc<Context>,
+    old_store: Arc<DhtStore<CAS, EAVS>>,
+    action_wrapper: &ActionWrapper,
+) -> Arc<DhtStore<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
+    EAVS: EntityAttributeValueStorage + Sized + Clone + PartialEq,
+{
+    // Get reducer
+    let maybe_reducer = resolve_reducer(action_wrapper);
+    if maybe_reducer.is_none() {
+        return old_store;
+    }
+    let reducer = maybe_reducer.unwrap();
+    // Reduce
+    let maybe_new_store = reducer(context, &old_store, &action_wrapper);
+    match maybe_new_store {
+        None => old_store,
+        Some(new_store) => Arc::new(new_store),
+    }
+}
+
+/// Maps incoming action to the correct reducer
+fn resolve_reducer<CAS, EAVS>(action_wrapper: &ActionWrapper) -> Option<DhtReducer<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
+    EAVS: EntityAttributeValueStorage + Sized + Clone + PartialEq,
+{
+    match action_wrapper.action() {
+        Action::Commit(_, _) => Some(reduce_commit_entry),
+        Action::GetEntry(_) => Some(reduce_get_entry),
+        Action::AddLink(_) => Some(reduce_add_link),
+        Action::GetLinks(_) => Some(reduce_get_links),
+        _ => None,
+    }
+}
+
+//
+pub(crate) fn reduce_commit_entry<CAS, EAVS>(
+    _context: Arc<Context>,
+    old_store: &DhtStore<CAS, EAVS>,
+    action_wrapper: &ActionWrapper,
+) -> Option<DhtStore<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
+    EAVS: EntityAttributeValueStorage + Sized + Clone + PartialEq,
+{
+    let action = action_wrapper.action();
+    let (_entry_type, entry) = match action {
+        Action::Commit(entry_type, entry) => (entry_type, entry),
+        _ => unreachable!(),
+    };
+    // Look in local storage if it already has it
+    if old_store
+        .content_storage()
+        .contains(&entry.address())
+        .unwrap()
+    {
+        // Maybe panic as this should never happen?
+        return None;
+    }
+    // Otherwise add it local storage...
+    let mut new_store = (*old_store).clone();
+    let res = new_store.content_storage_mut().add(entry);
+    if res.is_err() {
+        return None;
+    }
+    // ...and publish to the network
+    new_store.network_mut().publish(entry);
+    Some(new_store)
+}
+
+//
+pub(crate) fn reduce_get_entry<CAS, EAVS>(
```

Comment:
I thought that GetEntry only needed to be in the reducer if we know that we don't have it already, i.e. that we do the request for getting an entry directly from the state, and then it's only an action if we didn't find it locally because that action of getting it from the network is precisely the state that needs to be managed.  i.e. that there's more like a RequestEntryFromNetwork action which the reducer uses then to store the value so retrieved.

Reply:
Yes. So I guess we just need to rename the `GetEntry` Action to `GetEntryFromNetwork`, right @lucksus ?

Reply:
I would agree to rename it to something like that, or maybe LookupEntryInDht.
I think the reducer should be named this way already. You could also rename the action.

What is out of the scope of this PR is changing the agent/chain reducer for GetEntry and instead write an action creator..